libcontainer/process_linux.go

```diff
@@ -310,18 +311,95 @@ func (p *setnsProcess) addIntoCgroupV2() error {
 }
 
 func (p *setnsProcess) addIntoCgroup() error {
+	if p.cmd.SysProcAttr.UseCgroupFD {
+		return nil
+	}
 	if cgroups.IsCgroup2UnifiedMode() {
 		return p.addIntoCgroupV2()
 	}
 	return p.addIntoCgroupV1()
 }
 
+// prepareCloneIntoCgroup sets up p.cmd to use clone3 with CLONE_INTO_CGROUP
+// to join cgroup early. It returns an *os.File which must be closed by the
+// caller after p.Cmd.Start.
+func (p *setnsProcess) prepareCloneIntoCgroup() (*os.File, error) {
+	if !cgroups.IsCgroup2UnifiedMode() {
+		return nil, nil
+	}
+
+	base := p.manager.Path("")
+	if base == "" { // No cgroup to join.
+		return nil, nil
+	}
+	sub := ""
+	if p.process.SubCgroupPaths != nil {
+		sub = p.process.SubCgroupPaths[""]
+	}
+	cgroup := path.Join(base, sub)
+	if !strings.HasPrefix(cgroup, base) {
+		return nil, fmt.Errorf("bad sub cgroup path: %s", sub)
+	}
+
+	fd, err := os.OpenFile(cgroup, unix.O_PATH|unix.O_DIRECTORY|unix.O_CLOEXEC, 0)
+	if err != nil {
+		if p.rootlessCgroups {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("can't open cgroup: %w", err)
+	}
+
+	logrus.Debugf("using CLONE_INTO_CGROUP %q", cgroup)
+	if p.cmd.SysProcAttr == nil {
+		p.cmd.SysProcAttr = &syscall.SysProcAttr{}
+	}
+	p.cmd.SysProcAttr.UseCgroupFD = true
+	p.cmd.SysProcAttr.CgroupFD = int(fd.Fd())
+
+	return fd, nil
+}
+
+func (p *setnsProcess) shouldRetryWithoutCgroupFD(err error) bool {
+	switch {
+	case !p.cmd.SysProcAttr.UseCgroupFD:
+		// CgroupFD was not used.
+		return false
+	case errors.Is(err, errors.ErrUnsupported):
+		// clone3(CLONE_INTO_CGROUP) is not supported (ENOSYS),
+		// or cgroup is in the domain invalid state (EOPNOTSUPP).
+		return true
+	case errors.Is(err, unix.EBUSY):
+		// Nesting + domain controllers.
+		return true
+	case p.rootlessCgroups && errors.Is(err, unix.EACCES):
+		// Rootless has no direct access to cgroup.
+		return true
+	}
+
```

Comment:
I fear we might nor retry in a case we should and then the container will never start for the user and be a big regression. Can we log as debug what was the error before returning false?

So in case that happens, we can ask them to run with debug and check the error they get and we just handle that case?

Reply:
@kolyshkin thanks! Now with the debug log I feel more comfortable.

My concern is possible, though. Maybe I didn't express myself correctly.

We only retry without the cgroupfd if this function returns true. That is based on the manpage and the possible errors that are returned. It wouldn't be the first time a manpage is not up to date (i.e. clone_into_cgroup can fail with some other error too), nor the first time more errors than what we think possible are returned.

If this can fail with some other error than the ones handled here and it would succeed in that case when retrying without cgroupfd, then we will have a bug in that case.

This is the case I worry about.

Reply:
@kolyshkin what if we just always retry on a failure?

As other have pointed out, there were more errors we needed to handle. If we _always_ retry on failure without the cgroupfd, then it will:
 * Either work on retry, which is great
 * Or not work, in which case we just did one more try in the failing case (I wouldn't care that much about the failure to execute the container)

Then for sure we can not screw it up if the kernel in the future starts returning some other error or some red hat backport changes the error returned or whatever.

What do you think?